cupy/cuda/_ipc_handle.py

```diff
@@ -0,0 +1,109 @@
+import os
+import warnings
+from multiprocessing import get_start_method
+
+from cupy.cuda.memory import UnownedMemory, MemoryPointer
+
+from cupy.cuda.runtime import ipcGetMemHandle
+from cupy.cuda.runtime import ipcOpenMemHandle
+from cupy.cuda.runtime import ipcCloseMemHandle
+
+
+class IPCMemoryHandle:
+    """
+    A serializable class representing an Inter-process Communication (IPC)
+    handle for sharing CuPy arrays between processes.
+    """
+
+    def __init__(self, ndarray_instance):
+        """
+        Initialize an IPCMemoryHandle object.
+
+        Args:
+            ndarray_instance (cupy.ndarray): CuPy array instance to create
+            the handle.
+        """
+        from cupy._core.core import ndarray
+
+        if type(ndarray_instance) is not ndarray \
+                and isinstance(ndarray_instance, ndarray):
+            warnings.warn("The provided object is an instance of a "
+                          "subclass of cupy.ndarray, not cupy.ndarray "
+                          "itself. The .open() method will always return "
+                          "a cupy.ndarray object.")
+
+        if not isinstance(ndarray_instance, ndarray):
+            raise TypeError(
+                "The provided object is neither an instance of "
+                "cupy.ndarray, nor an instance of a subclass of "
+                "cupy.ndarray."
+            )
+
+        self.dtype = ndarray_instance.dtype
+        self.shape = ndarray_instance.shape
+        self.size = ndarray_instance.size
+        self.strides = ndarray_instance.strides
+        self.handle = ipcGetMemHandle(ndarray_instance.data.ptr)
+        self.opened = False
+        self.source_pid = os.getpid()
+        self.arr_ptr = -1
+
+    def __del__(self):
+        """
+        Closes the handle when object is deleted or goes out of scope.
+        """
+        self.close()
```

Comment:
Our experience shows that a bare `__del__` has terrible issues during Python interpreter shutdown. Please replace this with either [`is_shutting_down`](https://github.com/cupy/cupy/blob/03b45e5ce22599ce1f167a37124592660adf2620/cupy/cuda/stream.pyx#L130) ([preferred](https://github.com/cupy/cupy/pull/8315#issuecomment-3303982492)) or a [weakref finalizer](https://github.com/cupy/cupy/pull/8315).